plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs

```diff
@@ -521,6 +526,46 @@ instance (HasConstantIn uni term, GShow uni, GEq uni, uni `Contains` a) =>
 -- | A constraint for \"@a@ is a 'KnownType' by means of being included in @UniOf term@\".
 type KnownBuiltinType term a = KnownBuiltinTypeIn (UniOf term) term a
 
+{- Note [Performance of KnownTypeIn instances]
+Even though we don't use 'makeKnown' and 'readKnown' directly over concrete types, it's still
+beneficial to inline them, because otherwise GHC compiles each of them to two definitions
+(one calling the other) for some reason.
+
+Similarly, we inline implementations of 'knownTypeAst' just to get tidier Core.
+
+Some 'readKnown' implementations require inserting a call to 'oneShot'. E.g. if 'oneShot' is not
+used in 'readKnownConstant' then 'GHC pulls @gshow uniExp@ out of the 'Nothing' branch, thus
+allocating a thunk of type 'String' that is completely redundant whenever there's no error,
+which is the majority of cases. And putting 'oneShot' as the outermost call results in
+worse Core.
+
+Any change to an instance of 'KnownTypeIn', even completely trivial, requires looking into the
+generated Core, since compilation of these instances is extremely brittle optimization-wise.
+-}
+
+-- See Note [Performance of KnownTypeIn instances]
+-- | Convert a constant embedded into a PLC term to the corresponding Haskell value.
+readKnownConstant
+    :: forall term a err cause m.
+       ( MonadError (ErrorWithCause err cause) m, AsUnliftingError err
+       , KnownBuiltinType term a
+       )
+    => Maybe cause -> term -> m a
+readKnownConstant mayCause term = asConstant mayCause term >>= oneShot \case
+    Some (ValueOf uniAct x) -> do
```

Comment:
Why `uniAct`?  Does this mean the actual universe that we've got?

Reply:
Exactly. "act" = "actual", "exp" = "expected".